File: src/main/java/com/vc/medium/L114.scala
package com.vc.medium

class TreeNode114(var _value: Int) {
  var value: Int = _value
  var left: TreeNode114 = null
  var right: TreeNode114 = null
}

object L114 {
  def flatten(root: TreeNode114): Unit = {
    var prev: TreeNode114 = null
    def solve(root: TreeNode114): Unit = {
      if(root != null) {
        solve(root.right)
        solve(root.left)
        root.right = prev
        root.left = null
        prev = root
      }
    }
    solve(root)
    prev
  }
}

File: src/main/java/com/vc/medium/L114.java
package com.vc.medium;

import com.vc.hard.TreeNode;

class L114 {
    public void flatten(TreeNode root) {
        helper(root);
    }

    private TreeNode helper(TreeNode root) {
        if(root == null) return null;

        if(root.left == null && root.right == null) return root;

        TreeNode left = helper(root.left);
        TreeNode right = helper(root.right);

        if(left != null) {
            left.right = root.right;
            root.right = root.left;
            root.left = null;
        }
        return right == null ? left : right;
    }
}